openvpn/session/manager.go

```diff
@@ -1,36 +1,44 @@
 package session
 
 import (
+	"github.com/mysterium/node/identity"
 	"github.com/mysterium/node/openvpn"
 	"github.com/mysterium/node/session"
+	"sync"
 )
 
-func NewManager(clientConfig *openvpn.ClientConfig) *manager {
+//NewManager returns session manager which maintans a map of session id -> session
+func NewManager(clientConfig *openvpn.ClientConfig, idGenerator session.Generator) *manager {
 	return &manager{
-		idGenerator:  &session.Generator{},
+		idGenerator:  idGenerator,
 		clientConfig: clientConfig,
-		sessions:     make([]session.SessionID, 0),
+		sessionMap:   make(map[session.SessionID]session.Session),
+		creationLock: &sync.Mutex{},
 	}
 }
 
 type manager struct {
-	idGenerator  session.GeneratorInterface
+	idGenerator  session.Generator
 	clientConfig *openvpn.ClientConfig
-	sessions     []session.SessionID
+	sessionMap   map[session.SessionID]session.Session
+	creationLock *sync.Mutex
 }
 
-func (manager *manager) Create() (sessionInstance session.Session, err error) {
+func (manager *manager) Create(peerId identity.Identity) (sessionInstance session.Session, err error) {
+	manager.creationLock.Lock()
```

Comment:
why do we need this Lock?

Reply:
Because session manager method "create" can be accessed concurrently by multiple peer dialogs and map type is not thread safe in golang